lib/sqs.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+const assert = require('assert');
+const AWS    = require('aws-sdk');
+
+
+module.exports = class SQS {
+  constructor(config) {
+    this.accessKeyId     = config.accessKeyId;
+    this.secretAccessKey = config.secretAccessKey;
+    this.region          = config.region;
+    this.queueName       = config.queueName;
+  }
+
+  post(messages) {
+    return this.sqsClient()
+      .then(sqsClient => {
+        const entries = messages.map((message, index) => {
+          const entry = {
+            // Result may be unordered, that's why we need to provide an ID
+            // for each message in this batch.
+            Id:          index.toString(),
+            MessageBody: message.body
+          };
+
+          // This is the recommended way to check if a queue is FIFO or not.
+          if (this.queueName.endsWith('.fifo'))
+            entry.MessageGroupId = this.queueName;
+
+          return entry;
+        });
+
+        const sqsPayload = {
+          QueueUrl: this.queueURL,
+          Entries:  entries
+        };
+
+        return sqsClient.sendMessageBatch(sqsPayload).promise();
+      })
+      .then(response => {
+        const successful = response.Successful;
+        const failed     = response.Failed;
+
+        if (failed.length) {
+          const errors = failed.map(error => `${error.Code}: ${error.Message}`);
+          throw new Error(errors.join(', '));
+        }
+
+        // Reconstruct array of job IDs based on the batch ID for this request.
+        const orderedJobIDs = messages.map(function(message, index) {
+          const responseResult = successful.find(msg => msg.Id === index.toString());
+          assert(responseResult);
```

Comment:
But what's the assumption?

That we queued N messages but we only got confirmation for < N and no error indicating some messages were not queued? That would be easy to test by comparing the length, and would need an appropriate error message.

That SQS did queue N messages, but somehow forgot to include the message ID for them all. Maybe it has this bug, sounds like `assert(messages.every(....))` is the way to find out.

That SQS did queue N messages, but returned IDs for some other set of messages we didn't ask it to queue?